### ISSUE TITLE
ERROR: per_read_mods output requested, but specified model does not support calling modified bases

### ISSUE THREAD

**Issue by jon-xu** [nanoporetech/megalodon#280]
Hi Marcus,

We are trying to use megalodon and taiyaki to train the model with a sample with pseudouridine modification at known positions.
When we prepare the training data in megaladon before we train the model using taiyaki, we met the error as in the subject.

The command we used is:

megalodon $WDR/modified/fastq/workspace \
        --outputs mappings per_read_mods \
        --output-directory $WDR/training_prep \
        --reference $WDR/reference_sequence/2021-8_pBASE1_eGFP_sequence_2.fasta \
        --guppy-config $WDR/ont-guppy/data/rna_r9.4.1_70bps_hac.cfg \
        --guppy-server-path $WDR/ont-guppy/bin/guppy_basecall_server \
        --rna --guppy-config rna_r9.4.1_70bps_hac.cfg \
        --devices 0,1 \
        --processes 40

As I understand, there's no modification-ready models for RNA, and we are handling pseudouridine which should be new.

Do you have any advice how we should proceed, please?
Thanks,
Jon

**Comment by marcus1487** [comment]
Your command line argument includes two different guppy configs. This is likely the cause of your issue here.

**Comment by jon-xu** [comment]
Thanks Marcus, I just removed the first guppy-config and the error message kept the same.

**Comment by marcus1487** [comment]
Could you explain a bit more on your goal from this megalodon run? Are you trying to produce training data for a modified base model? Or trying to call modified bases with a model you have trained to detect pseU? if it is the former then per_read_mods are not available from the standard model. We do not currently have a released model to detect RNA modified bases.

**Comment by jon-xu** [comment]
We have a modified RNA (all Uridine modified) sample with another unmodified sample as reference for the location of U's.
And we want to train a model which can detect pseU in other samples.

So we are following the section "Ground Truth Aided Bootstrap Modified Base Annotation" on https://nanoporetech.github.io/megalodon/modbase_training.html#ground-truth-aided-bootstrap-modified-base-annotation

To do that, right before the megalodon_extras command, the documentation said:
"Given a completed Megalodon run (megalodon_results with --outputs mappings per_read_mods) and the ground truth bedmethyl file (ground_truth_methylation.CG.bed) the following command will compute per-site modified base thresholds and low coverage sites."

That's why we were running the previously mentioned command.

Please kindly advise what the correct steps we should take.
Many thanks!

**Comment by marcus1487** [comment]
Ground truth aided training requires a model which calls that modified base as a starting point. The aim of this pipeline is to improve an existing model. 

Your sample will follow the motif pipeline but with the motif being “U 0”. 

I would warn that a sample with completely modified bases is unlikely to work well on partially methylated data. If your goal is to discriminate completely modified strands from canonical strands then this model will work well. Call sparse pseU in a “native” sample is unlikely to work well from the training data you have described. 

Appropriate training samples for calling realistic/native data is much more complex and an active area of internal research.

**Comment by jon-xu** [comment]
Many thanks Marcus for the advice!

**Comment by jon-xu** [comment]
Hi Marcus,
When we are trying the motif approach, we met an error:
"AssertionError: All bases in collapse alphabet must occur within alphabet."
Shall we use T 0 instead? It's an RNA sample.
Thanks!

**Comment by marcus1487** [comment]
Yes. Megalodon generally converts Us to Ts for RNA processing. I'd have to have a look at this specific case, but probably best to just work with T even for RNA data from the informatics side.

**Comment by jon-xu** [comment]
Thanks for your reply Marcus!
I followed the motif-base pipeline and got the model trained successfully.
But when I basecall using the new model on the same data I used for training, most of the reads went fail...

**Comment by marcus1487** [comment]
Basecalling models need calibration for the q-scores (not sure this code it robust or public). I'd recommend just turning off any q-score filtering.

**Comment by jon-xu** [comment]
Hi Marcus, now I have finished calibration and generated "megalodon_mod_calibration.npz".
And next I would like to run megalodon again with the taiyaki-trained model and with the calibration file. But I met an error:
"

Compared to last successful run, I just changed --disable-mod-calibration to --mod-calibration-filename $WDR/megalodon_calibrated/mod_calibration_statistics.npz:
 
megalodon $WDR/data/drna/20220321_VAC_mU_fast5 \
        --outputs basecalls mappings mod_mappings mods per_read_mods \
        --output-directory $WDR/megalodon_validate/20220321_VAC_mU_fast5 \
        --reference $WDR/reference_sequence/2021-8_pBASE1_eGFP_sequence_2.fasta \
        --guppy-server-path $WDR/ont-guppy/bin/guppy_basecall_server \
        --rna --guppy-config rna_r9.4.1_70bps_hac_mU.cfg \
        --mod-calibration-filename $WDR/megalodon_calibrated/mod_calibration_statistics.npz \
        --devices 0,1 \
        --processes 40 --overwrite

Please kindly advise.
Thanks,
Jon

**Comment by jon-xu** [comment]
Hi Marcus,
Do you have some suggestion on this, please?
It worked all well with --disable-mod-calibration and only happened when I used --mod-calibration-filename.
Many thanks!
Jon
